### ISSUE TITLE
Prerendering

### ISSUE THREAD

**Issue by danroth27** [aspnet/AspNetCore#5464]
Also known as Server-Side Rendering (SSR)

**Comment by andras-ferencz** [comment]
For anyone who gets blocked by the lack of serverside rendering, this is a class I just wrote to resolve the issue, which was preventing me from being crawled by AdSense.

I have confirmed that it works (using the Google Search Console) and does add some, but not a significant amount overhead to crawling response times. It might be better to do something similar but prerender all of the content and store it in a folder and serve it upon request by a crawler - I dunno. 

It is also faster if you change HostURI to the localhost.


It supports all known crawlers inside the JSON data file from https://github.com/monperrus/crawler-user-agents.

It depends on the NuGet package **Selenium.WebDriver** v3.141.0 and you also **must have Chrome installed on the target machine**.

Usage goes something like this where you set the `MagicWord` string to something you expect to exist in the HTML after rendering is complete (such as a header or footer ID found in the rendered Blazor components).

       public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            SeleniumRenderMiddleware.MagicWord = "language_bar";
            app.UseMiddleware<SeleniumRenderMiddleware>();
            app.UseServerSideBlazor<App.Startup>();
        }


```C#
using System.IO;
using System.Net;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using System.IO.Compression;
using System.Runtime.InteropServices;
using System;
using System.Collections.Generic;
using Microsoft.JSInterop;
using OpenQA.Selenium.Chrome; //Nuget Package: Selenium.WebDriver v3.141.0
using System.Reflection;

namespace SeleniumRender
{
    public class UserAgents
    {
        public string pattern { get; set; }
        public string url { get; set; }
        public string[] instances { get; set; }
        public string addition_date { get; set; }
        public string description { get; set; }
    }

    public class SeleniumRenderMiddleware
    {
        // GOOGLE CHROME MUST BE PRE-INSTALLED ON THE TARGET MACHINE FOR THIS TO WORK.
        private static Dictionary<string, string> s_userAgents = new Dictionary<string, string>();
        private static ChromeOptions s_chromeOptions = new ChromeOptions() { AcceptInsecureCertificates = true };
        private static ChromeDriver s_chromeDriver;
        private static string s_chromeDriverFilename
        {
            get
            {
                if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) return "chromedriver_win32.zip	";
                else if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) return "chromedriver_linux64.zip";
                else if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) return "chromedriver_mac64.zip";
                throw new Exception("Are you running OS/2 Warp?!");
            }
        }
        private static string s_chromeVersion = "73.0.3683.20";
        public static string HostURI;
        public static string MagicWord = "0xDEADBEEF";
        private RequestDelegate _next;
        static SeleniumRenderMiddleware()
        {
            string ChromeDriverLocation = Path.GetDirectoryName(Assembly.GetEntryAssembly().Location); 

            if (Directory.Exists(ChromeDriverLocation) == false)
                Directory.CreateDirectory(ChromeDriverLocation);

            using (var client = new WebClient())
            {
                    if (File.Exists($"{ChromeDriverLocation}\\agents.json") == false)
                        client.DownloadFile("https://raw.githubusercontent.com/monperrus/crawler-user-agents/master/crawler-user-agents.json", $"{ChromeDriverLocation}\\agents.json");
                    if (File.Exists($"{ChromeDriverLocation}\\chromedriver.exe") == false)
                    {
                        byte[] zipFile = client.DownloadData($"https://chromedriver.storage.googleapis.com/{s_chromeVersion}/{s_chromeDriverFilename}");
                        using (var ms = new MemoryStream(zipFile))
                        {
                            ms.Seek(0, SeekOrigin.Begin);
                            ZipArchive archive = new ZipArchive(ms);
                            archive.ExtractToDirectory(ChromeDriverLocation);
                        }
                    }
            }

            foreach (var userAgent in Json.Deserialize<UserAgents[]>(File.ReadAllText($"{ChromeDriverLocation}\\agents.json")))
                foreach (var instance in userAgent.instances)
                    s_userAgents.Add(instance, userAgent.pattern);

            s_chromeOptions.AddArgument("headless");
            s_chromeDriver = new ChromeDriver(ChromeDriverLocation, s_chromeOptions);
        }
        public SeleniumRenderMiddleware(RequestDelegate next) => _next = next;
        public async Task Invoke(HttpContext context)
        {
            if (s_userAgents.ContainsKey(context.Request.Headers["User-Agent"]) == false)
                await _next.Invoke(context);
            else
            {
                if (HostURI == null)
                    HostURI = (context.Request.IsHttps ? "https://" : "http://") + context.Request.Host.Value;
                s_chromeDriver.Url = HostURI + context.Request.Path;
                s_chromeDriver.Navigate();

                while (s_chromeDriver.PageSource.Contains(MagicWord) == false)
                    await Task.Delay(100);

                context.Response.ContentType = "text/html";
                await context.Response.WriteAsync(s_chromeDriver.PageSource);
            }
        }
    }
}`

**Comment by javiercn** [comment]
Prerrendering is supported since preview2. 

See https://blogs.msdn.microsoft.com/webdev/2019/01/29/aspnet-core-3-preview-2/

The part Integration with MVC Views and Razor Pages

**Comment by andras-ferencz** [comment]
If I'm not mistaken, this can't presently be used in a middle-ware layer to serve crawlers static pages, correct?

**Comment by ElectricHavoc** [comment]
_**TLDR; +1 for 98% C# SPA that does full render on any path from refresh.**_

> Components rendered from pages and views will be prerendered, but are not yet interactive (i.e. clicking the Counter button doesn't do anything in this release). 

Non-interactive is not support (imo) it confuses people and probably should have been baked more. I've also been searching for a sample to get this working, to no avail. 

I hope this applies to not only components, but mini applications:
```
@(await Html.RenderComponentAsync<Blazor.Applications.LoginApp>())
@(await Html.RenderComponentAsync<Blazor.Applications.CartApp>())
@(await Html.RenderComponentAsync<Blazor.Applications.ProductsApp>())
```

Or perhaps better syntax for this usage:
```
@(await Html.RenderComponentAppAsync<Blazor.Applications.ProductsApp>())
```
Which would still handle some form of child url routing from where it's rendered.

__Example__
If ProductsApp (RazorComponentApp) has routes "/" & "/specials"
If this was placed on a "/page" mvc/razor route "/page/specials" would still resolve, prerender and serve from the ProductsApp component from where it was called in the application. _(Hopefully the idea is articulated here without too much detail)_

The lack of ability to have a 100% rendered page on direct browse to /counter is what I believe to be the limiting factor on most corporate/commercial projects. Especially ones that do not want either the full framework or entire application sent downstream. _(Not everyone is making a mobile spa or electron app)_

Of course, if a Blazor app could do full prerendering, I think all this is moot and we would convert our apps to full on server Blazor, and I second the suggestion that some areas do not need to be prerendered and would love `<Counter @prerender/> and/or @page(prerender: true) "/counter"` or similar for optimization.

_My 2c, is that my apps are far from SPA for various reasons, both functional and SEO are huge considerations. I think this is an amazing use of C# that I am looking forward to see expanded. This is what `<asp:UpdatePanel>` should have been all along, with the ability to route._ 

Also, to those asking why you would NOT prerender a portion on a prerendered whole: You may need to have calls to external services you don't want the server side making... calls to weather.json for example, the server would prerender the site, SEO is all good, and then collects data from potentially _other_ servers/services.

**Comment by javiercn** [comment]
This has been done as part of https://github.com/aspnet/AspNetCore/pull/7770
There are some minor improvements left that will be tracked separately.
